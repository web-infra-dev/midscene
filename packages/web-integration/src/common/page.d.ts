import type { KeyInput } from 'puppeteer';
import type { AppiumPage } from '../appium';
import type { PlaywrightWebPage } from '../playwright';
import type { PuppeteerWebPage } from '../puppeteer';

export type WebPage = PlaywrightWebPage | PuppeteerWebPage | AppiumPage;
export type WebKeyInput = KeyInput;
