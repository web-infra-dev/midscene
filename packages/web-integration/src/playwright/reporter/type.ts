import type { Location } from '@playwright/test/reporter';

export type TestData = {
  testId: string;
  title: string;
  status: 'passed' | 'failed' | 'timedOut' | 'skipped' | 'interrupted';
  /**
   * Running time in milliseconds.
   */
  duration: number;
  /**
   * Optional location in the source where the step is defined.
   */
  location?: Location;
  dumpPath?: string;
};
