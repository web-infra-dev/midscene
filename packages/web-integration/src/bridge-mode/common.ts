export const DefaultBridgeServerPort = 3766;
export const DefaultLocalEndpoint = `http://127.0.0.1:${DefaultBridgeServerPort}`;
export const BridgeCallTimeout = 30000;

export enum BridgeEvent {
  Call = 'bridge-call',
  CallResponse = 'bridge-call-response',
  UpdateAgentStatus = 'bridge-update-agent-status',
  Message = 'bridge-message',
  Connected = 'bridge-connected',
  Refused = 'bridge-refused',
  ConnectNewTabWithUrl = 'connectNewTabWithUrl',
  ConnectCurrentTab = 'connectCurrentTab',
}

export interface BridgeConnectTabOptions {
  /**
   * If true, the page will always track the active tab.
   * @default true
   */
  forceSameTabNavigation?: boolean;
}

export enum MouseEvent {
  PREFIX = 'mouse.',
  Click = 'mouse.click',
  Wheel = 'mouse.wheel',
  Move = 'mouse.move',
  Drag = 'mouse.drag',
}

export enum KeyboardEvent {
  PREFIX = 'keyboard.',
  Type = 'keyboard.type',
  Press = 'keyboard.press',
}

export const BridgePageType = 'page-over-chrome-extension-bridge';

export const BridgeErrorCodeNoClientConnected = 'no-client-connected';

export interface BridgeCall {
  method: string;
  args: any[];
  response: any;
  callTime: number;
  responseTime: number;
  callback: (error: Error | undefined, response: any) => void;
  error?: Error;
}

export interface BridgeCallRequest {
  id: string;
  method: string;
  args: any[];
}

export interface BridgeCallResponse {
  id: string;
  response: any;
  error?: any;
}

export interface BridgeConnectedEventPayload {
  version: string;
}
