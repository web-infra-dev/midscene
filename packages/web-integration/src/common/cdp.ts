// fork from https://github.com/zerostep-ai/zerostep/blob/5ba8ca5282879f444e5dfefbcc1f03c76591469e/packages/playwright/src/cdp.ts
import type { Page } from 'playwright';

const cdpSessionByPage = new Map<Page, CDPSession>();

export const WEBDRIVER_ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf';

export type ScrollType = 'up' | 'down' | 'bottom' | 'top';

/**
 * Closes the cdp session and clears the global shared reference. This
 * happens automatically when a page closes in a playwright test, so
 * should generally not be necessary.
 */
export const detachCPDSession = async (page: Page) => {
  if (cdpSessionByPage.has(page)) {
    await cdpSessionByPage.get(page)!.detach();
    cdpSessionByPage.delete(page);
  }
};

/**
 * Returns a stable reference to a CDP session.
 */
export const getCDPSession = async (page: Page): Promise<CDPSession> => {
  if (!cdpSessionByPage.has(page)) {
    try {
      const session = await page.context().newCDPSession(page);
      cdpSessionByPage.set(page, session);
    } catch (e) {
      throw Error('The ai() function can only be run against Chromium browsers.');
    }
  }

  return cdpSessionByPage.get(page)!;
};

export const getScreenshot = async (page: Page) => {
  const cdpSession = await getCDPSession(page);
  const screenshot = await cdpSession.send('Page.captureScreenshot');
  return screenshot.data; // Base64-encoded image data
};

export const scrollIntoView = async (page: Page, args: { id: string }) => {
  const cdpSession = await getCDPSession(page);

  await cdpSession.send('DOM.scrollIntoViewIfNeeded', {
    backendNodeId: parseInt(args.id, 10),
  });
};

export const getTitle = async (page: Page) => {
  const cdpSession = await getCDPSession(page);
  const returnedValue = await cdpSession.send('Runtime.evaluate', {
    expression: 'document.title',
    returnByValue: true,
  });

  return returnedValue.result.value;
};

export const get = async (page: Page, args: { url: string }) => {
  const cdpSession = await getCDPSession(page);
  await cdpSession.send('Page.navigate', {
    url: args.url,
  });
};

export const scrollElement = async (page: Page, args: { id: string; target: ScrollType }) => {
  await runFunctionOn(page, {
    functionDeclaration: `function() {
      let element = this
      let elementHeight = 0

      switch (element.tagName) {
        case 'BODY':
        case 'HTML':
          element = document.scrollingElement || document.body
          elementHeight = window.visualViewport?.height ?? 720
          break
        default:
          elementHeight = element.clientHeight ?? 720
          break
      }

      const relativeScrollDistance = 0.75 * elementHeight

      switch ("${args.target}") {
        case 'top':
          return element.scrollTo({ top: 0 })
        case 'bottom':
          return element.scrollTo({ top: element.scrollHeight })
        case 'up':
          return element.scrollBy({ top: -relativeScrollDistance })
        case 'down':
          return element.scrollBy({ top: relativeScrollDistance })
        default:
          throw Error('Unsupported scroll target ${args.target}')
      }
    }`,
    backendNodeId: parseInt(args.id, 10),
  });
};

export const runFunctionOn = async (
  page: Page,
  args: { functionDeclaration: string; backendNodeId: number },
) => {
  const cdpSession = await getCDPSession(page);
  const {
    object: { objectId },
  } = await cdpSession.send('DOM.resolveNode', { backendNodeId: args.backendNodeId });
  await cdpSession.send('Runtime.callFunctionOn', {
    functionDeclaration: args.functionDeclaration,
    objectId,
  });
};

export const clearElement = async (page: Page, args: { id: string }) => {
  return await runFunctionOn(page, {
    functionDeclaration: `function() {this.value=''}`,
    backendNodeId: parseInt(args.id, 10),
  });
};

export const sendKeysToElement = async (page: Page, args: { id: string; value: string[] }) => {
  const cdpSession = await getCDPSession(page);
  const value = args.value[0];

  const { nodeId } = await cdpSession.send('DOM.requestNode', { objectId: args.id });
  await cdpSession.send('DOM.focus', { nodeId });

  // eslint-disable-next-line @typescript-eslint/prefer-for-of
  for (let i = 0; i < value.length; i++) {
    await cdpSession.send('Input.dispatchKeyEvent', {
      type: 'char',
      text: value[i],
    });
  }

  return true;
};

export const getElementAttribute = async (page: Page, args: { id: string; name: string }) => {
  const cdpSession = await getCDPSession(page);

  const { nodeId } = await cdpSession.send('DOM.requestNode', { objectId: args.id });
  const { attributes } = await cdpSession.send('DOM.getAttributes', { nodeId });

  for (let i = 0; i < attributes.length; i++) {
    if (attributes[i] === args.name) {
      return attributes[i + 1];
    }
  }
  return attributes;
};

export const getElementTagName = async (page: Page, args: { id: string }) => {
  const cdpSession = await getCDPSession(page);
  const returnedValue = await cdpSession.send('Runtime.callFunctionOn', {
    functionDeclaration: `function() {return this.tagName}`,
    objectId: args.id,
    returnByValue: true,
  });

  return returnedValue.result.value;
};

export const clickElement = async (page: Page, args: { id: string }) => {
  const cdpSession = await getCDPSession(page);
  const { centerX, centerY } = await getContentQuads(page, { backendNodeId: parseInt(args.id, 10) });

  await cdpSession.send('Input.dispatchMouseEvent', {
    type: 'mousePressed',
    x: centerX,
    y: centerY,
    button: 'left',
    clickCount: 1,
    buttons: 1,
  });

  await cdpSession.send('Input.dispatchMouseEvent', {
    type: 'mouseReleased',
    x: centerX,
    y: centerY,
    button: 'left',
    clickCount: 1,
    buttons: 1,
  });

  return true;
};

export const getContentQuads = async (page: Page, args: { backendNodeId: number }) => {
  const cdpSession = await getCDPSession(page);
  const quadsResponse = await cdpSession.send('DOM.getContentQuads', args);

  const [topLeftX, topLeftY, topRightX, topRightY, bottomRightX, bottomRightY, bottomLeftX, bottomLeftY] =
    quadsResponse.quads[0];

  const width = topRightX - topLeftX;
  const height = bottomRightY - topRightY;
  const centerX = topLeftX + width / 2;
  const centerY = topRightY + height / 2;

  console.log('getContentQuads', args, {
    width,
    height,
    topLeftX,
    topLeftY,
    centerX,
    centerY,
  });
  return {
    topLeftX,
    topLeftY,
    topRightX,
    topRightY,
    bottomRightX,
    bottomRightY,
    bottomLeftX,
    bottomLeftY,
    width,
    height,
    centerX,
    centerY,
  };
};

export const focusElement = async (page: Page, args: { backendNodeId: number }) => {
  const cdpSession = await getCDPSession(page);
  await cdpSession.send('DOM.focus', args);
};

export const getElementRect = async (page: Page, args: { id: string }) => {
  const cdpSession = await getCDPSession(page);
  const returnedValue = await cdpSession.send('Runtime.callFunctionOn', {
    functionDeclaration: `function() {return JSON.parse(JSON.stringify(this.getBoundingClientRect()))}`,
    objectId: args.id,
    returnByValue: true,
  });

  return returnedValue.result.value;
};

export const findElements = async (page: Page, args: { using: string; value: string }) => {
  switch (args.using) {
    case 'css selector':
    case 'tag name':
      return await querySelectorAll(page, { selector: args.value });
    default:
      throw Error(`Unsupported findElements strategy ${args.using}`);
  }
};

export const querySelectorAll = async (page: Page, args: { selector: string }) => {
  const cdpSession = await getCDPSession(page);
  const rootDocumentNode = await cdpSession.send('DOM.getDocument', { depth: -1 });
  const returned = await cdpSession.send('DOM.querySelectorAll', {
    nodeId: rootDocumentNode.root.nodeId,
    selector: args.selector,
  });
  const resolvedNodesPromises = returned.nodeIds.map(
    async (nodeId) => await cdpSession.send('DOM.resolveNode', { nodeId }),
  );
  const resolvedNodes = await Promise.all(resolvedNodesPromises);
  const returnValue = resolvedNodes.map((node) => ({ [WEBDRIVER_ELEMENT_KEY]: node.object.objectId }));
  return returnValue;
};

export const getCurrentUrl = async (page: Page) => {
  const cdpSession = await getCDPSession(page);
  const returned = await cdpSession.send('Page.getNavigationHistory');
  const returnValue = returned.entries[returned.currentIndex].url;
  return returnValue;
};

export const executeScript = async (page: Page, args: { script: string; args: any[] }) => {
  const functionDeclaration = `function() { ${args.script} }`;
  const functionArgs = args.args.map((arg) => {
    if (typeof arg === 'boolean' || typeof arg === 'string' || typeof arg === 'number') {
      return { value: arg };
    } else if (arg && typeof arg === 'object' && Reflect.has(arg, WEBDRIVER_ELEMENT_KEY)) {
      return { objectId: arg[WEBDRIVER_ELEMENT_KEY] };
    } else {
      return { value: undefined };
    }
  });

  const cdpSession = await getCDPSession(page);
  await cdpSession.send('Runtime.enable');
  const window = await cdpSession.send('Runtime.evaluate', { expression: 'window' });

  const returnedRef = await cdpSession.send('Runtime.callFunctionOn', {
    objectId: window.result.objectId,
    functionDeclaration,
    arguments: functionArgs,
  });

  if (returnedRef.result.className === 'NodeList') {
    const nodeProperties = await cdpSession.send('Runtime.getProperties', {
      objectId: returnedRef.result.objectId!,
      ownProperties: true,
    });
    return nodeProperties.result
      .map((e: any) => (!isNaN(parseInt(e.name, 10)) ? { [WEBDRIVER_ELEMENT_KEY]: e.value?.objectId } : null))
      .filter((e: any) => e);
  } else if (returnedRef.result.className === 'HTMLHtmlElement') {
    return { [WEBDRIVER_ELEMENT_KEY]: returnedRef.result.objectId };
  } else {
    const returnedValue = await cdpSession.send('Runtime.callFunctionOn', {
      objectId: window.result.objectId,
      functionDeclaration,
      arguments: functionArgs,
      returnByValue: true,
    });

    return returnedValue.result.value;
  }
};

type CDPSession = Awaited<ReturnType<ReturnType<Page['context']>['newCDPSession']>>;
