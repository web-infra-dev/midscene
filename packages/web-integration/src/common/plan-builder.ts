import type {
  DetailedLocateParam,
  PlanningAction,
  PlanningActionParamHover,
  PlanningActionParamInputOrKeyPress,
  PlanningActionParamScroll,
  PlanningActionParamSleep,
  PlanningActionParamTap,
  PlanningLocateParam,
} from '@midscene/core';
import { getDebug } from '@midscene/shared/logger';
import { assert } from '@midscene/shared/utils';

const debug = getDebug('plan-builder');

export function buildPlans(
  type: PlanningAction['type'],
  locateParam?: DetailedLocateParam,
  param?:
    | PlanningActionParamInputOrKeyPress
    | PlanningActionParamScroll
    | PlanningActionParamSleep,
): PlanningAction[] {
  let returnPlans: PlanningAction[] = [];
  const locatePlan: PlanningAction<PlanningLocateParam> | null = locateParam
    ? {
        type: 'Locate',
        locate: locateParam,
        param: locateParam,
        thought: '',
      }
    : null;
  if (type === 'Tap' || type === 'Hover') {
    assert(locateParam, `missing locate info for action "${type}"`);
    assert(locatePlan, `missing locate info for action "${type}"`);
    const tapPlan: PlanningAction<PlanningActionParamTap> = {
      type,
      param: null,
      thought: '',
      locate: locateParam,
    };

    returnPlans = [locatePlan, tapPlan];
  }
  if (type === 'Input' || type === 'KeyboardPress') {
    if (type === 'Input') {
      assert(locateParam, `missing locate info for action "${type}"`);
    }
    assert(param, `missing param for action "${type}"`);

    const inputPlan: PlanningAction<PlanningActionParamInputOrKeyPress> = {
      type,
      param: param as PlanningActionParamInputOrKeyPress,
      thought: '',
      locate: locateParam!,
    };

    if (locatePlan) {
      returnPlans = [locatePlan, inputPlan];
    } else {
      returnPlans = [inputPlan];
    }
  }

  if (type === 'Scroll') {
    assert(param, `missing param for action "${type}"`);

    const scrollPlan: PlanningAction<PlanningActionParamScroll> = {
      type,
      param: param as PlanningActionParamScroll,
      thought: '',
      locate: locateParam,
    };

    if (locatePlan) {
      returnPlans = [locatePlan, scrollPlan];
    } else {
      returnPlans = [scrollPlan];
    }
  }

  if (type === 'Sleep') {
    assert(param, `missing param for action "${type}"`);

    const sleepPlan: PlanningAction<PlanningActionParamSleep> = {
      type,
      param: param as PlanningActionParamSleep,
      thought: '',
      locate: null,
    };

    returnPlans = [sleepPlan];
  }

  if (returnPlans) {
    debug('buildPlans', returnPlans);
    return returnPlans;
  }

  throw new Error(`Not supported type: ${type}`);
}
