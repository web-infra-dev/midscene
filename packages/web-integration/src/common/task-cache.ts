import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import type { AIElementParseResponse, PlanningAction } from '@midscene/core';
import {
  getLogDirByType,
  stringifyDumpData,
  writeLogFile,
} from '@midscene/core/utils';
import { getMidscenePkgInfo } from '@midscene/shared/fs';
import { type WebUIContext, generateCacheId } from './utils';

export type PlanTask = {
  type: 'plan';
  prompt: string;
  pageContext: {
    url: string;
    size: {
      width: number;
      height: number;
    };
  };
  response: { plans: PlanningAction[] };
};

export type LocateTask = {
  type: 'locate';
  prompt: string;
  pageContext: {
    url: string;
    size: {
      width: number;
      height: number;
    };
  };
  response: AIElementParseResponse;
};

export type AiTasks = Array<PlanTask | LocateTask>;

export type AiTaskCache = {
  aiTasks: Array<{
    prompt: string;
    tasks: AiTasks;
  }>;
};

export class TaskCache {
  cache: AiTaskCache;

  cacheId: string;

  newCache: AiTaskCache;

  midscenePkgInfo: ReturnType<typeof getMidscenePkgInfo>;

  constructor(opts?: { fileName?: string }) {
    this.midscenePkgInfo = getMidscenePkgInfo(__dirname);
    this.cacheId = generateCacheId(opts?.fileName);
    this.cache = this.readCacheFromFile() || {
      aiTasks: [],
    };
    this.newCache = {
      aiTasks: [],
    };
  }

  getCacheGroupByPrompt(aiActionPrompt: string) {
    const { aiTasks = [] } = this.cache || { aiTasks: [] };
    const index = aiTasks.findIndex((item) => item.prompt === aiActionPrompt);
    const newCacheGroup: AiTasks = [];
    this.newCache.aiTasks.push({
      prompt: aiActionPrompt,
      tasks: newCacheGroup,
    });
    return {
      readCache: <T extends 'plan' | 'locate'>(
        pageContext: WebUIContext,
        type: T,
        actionPrompt: string,
      ) => {
        if (index === -1) {
          return false;
        }
        if (type === 'plan') {
          return this.readCache(
            pageContext,
            type,
            actionPrompt,
            aiTasks[index].tasks,
          ) as T extends 'plan' ? PlanTask['response'] : LocateTask['response'];
        }
        return this.readCache(
          pageContext,
          type,
          actionPrompt,
          aiTasks[index].tasks,
        ) as T extends 'plan' ? PlanTask['response'] : LocateTask['response'];
      },
      saveCache: (cache: PlanTask | LocateTask) => {
        newCacheGroup.push(cache);
        this.writeCacheToFile();
      },
    };
  }

  /**
   * Read and return cached responses asynchronously based on specific criteria
   * This function is mainly used to read cached responses from a certain storage medium.
   * It accepts three parameters: the page context information, the task type, and the user's prompt information.
   * In the function, it first checks whether there is cached data. If there is, it retrieves the first task response from the cache.
   * It then checks whether the task type is 'locate' and whether the corresponding element can be found in the new context.
   * If the element cannot be found, it returns false, indicating that the cache is invalid.
   * If the task type is correct and the user prompt matches, it checks whether the page context is the same.
   * If the page context is the same, it returns the cached response, indicating that the cache hit is successful.
   * If there is no cached data or the conditions are not met, the function returns false, indicating that no cache is available or the cache is not hit.
   *
   * @param pageContext UIContext<WebElementInfo> type, representing the context information of the current page
   * @param type String type, specifying the task type, can be 'plan' or 'locate'
   * @param userPrompt String type, representing user prompt information
   * @return Returns a Promise object that resolves to a boolean or object
   */
  readCache(
    pageContext: WebUIContext,
    type: 'plan',
    userPrompt: string,
    cacheGroup: AiTasks,
  ): PlanTask['response'];
  readCache(
    pageContext: WebUIContext,
    type: 'locate',
    userPrompt: string,
    cacheGroup: AiTasks,
  ): LocateTask['response'];
  readCache(
    pageContext: WebUIContext,
    type: 'plan' | 'locate',
    userPrompt: string,
    cacheGroup: AiTasks,
  ): PlanTask['response'] | LocateTask['response'] | false {
    if (cacheGroup.length > 0) {
      const index = cacheGroup.findIndex((item) => item.prompt === userPrompt);

      if (index === -1) {
        return false;
      }

      const taskRes = cacheGroup.splice(index, 1)[0];

      // The corresponding element cannot be found in the new context
      if (
        taskRes?.type === 'locate' &&
        !taskRes.response?.elements.every((element) => {
          const findIndex = pageContext.content.findIndex(
            (contentElement) => contentElement.id === element.id,
          );
          if (findIndex === -1) {
            return false;
          }
          return true;
        })
      ) {
        return false;
      }
      if (
        taskRes &&
        taskRes.type === type &&
        taskRes.prompt === userPrompt &&
        this.pageContextEqual(taskRes.pageContext, pageContext)
      ) {
        return taskRes.response;
      }
    }
    return false;
  }

  pageContextEqual(
    taskPageContext: LocateTask['pageContext'],
    pageContext: WebUIContext,
  ) {
    return (
      taskPageContext.size.width === pageContext.size.width &&
      taskPageContext.size.height === pageContext.size.height
    );
  }

  /**
   * Generate task cache data.
   * This method is mainly used to create or obtain some cached data for tasks, and it returns a new cache object.
   * In the cache object, it may contain task-related information, states, or other necessary data.
   * It is assumed that the `newCache` property already exists in the current class or object and is a data structure used to store task cache.
   * @returns {Object} Returns a new cache object, which may include task cache data.
   */
  generateTaskCache() {
    return this.newCache;
  }

  readCacheFromFile() {
    const cacheFile = join(getLogDirByType('cache'), `${this.cacheId}.json`);
    if (process.env.MIDSCENE_CACHE === 'true' && existsSync(cacheFile)) {
      try {
        const data = readFileSync(cacheFile, 'utf8');
        const jsonData = JSON.parse(data);
        if (
          jsonData.pkgName !== this.midscenePkgInfo.name ||
          jsonData.pkgVersion !== this.midscenePkgInfo.version
        ) {
          return undefined;
        }
        return jsonData as AiTaskCache;
      } catch (err) {
        return undefined;
      }
    }
    return undefined;
  }

  writeCacheToFile() {
    const midscenePkgInfo = getMidscenePkgInfo(__dirname);
    writeLogFile({
      fileName: `${this.cacheId}`,
      fileExt: 'json',
      fileContent: stringifyDumpData(
        {
          pkgName: midscenePkgInfo.name,
          pkgVersion: midscenePkgInfo.version,
          cacheId: this.cacheId,
          ...this.newCache,
        },
        2,
      ),
      type: 'cache',
    });
  }
}
