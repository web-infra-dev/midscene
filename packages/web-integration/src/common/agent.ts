import type { WebPage } from '@/common/page';
import type {
  AgentWaitForOpt,
  ExecutionDump,
  GroupedActionDump,
} from '@midscene/core';
import {
  groupedActionDumpFileExt,
  stringifyDumpData,
  writeLogFile,
} from '@midscene/core/utils';
import { PageTaskExecutor } from '../common/tasks';
import type { AiTaskCache } from './task-cache';
import { printReportMsg, reportFileName } from './utils';

export interface PageAgentOpt {
  testId?: string;
  cacheId?: string;
  groupName?: string;
  groupDescription?: string;
  cache?: AiTaskCache;
  /* if auto generate report, default true */
  generateReport?: boolean;
  /* if auto print report msg, default true */
  autoPrintReportMsg?: boolean;
}

export class PageAgent {
  page: WebPage;

  dump: GroupedActionDump;

  reportFile?: string;

  reportFileName?: string;

  taskExecutor: PageTaskExecutor;

  opts: PageAgentOpt;

  constructor(page: WebPage, opts?: PageAgentOpt) {
    this.page = page;
    this.opts = Object.assign(
      {
        generateReport: true,
        autoPrintReportMsg: true,
        groupName: 'Midscene Report',
        groupDescription: '',
      },
      opts || {},
    );
    this.dump = {
      groupName: this.opts.groupName!,
      groupDescription: this.opts.groupDescription,
      executions: [],
    };
    this.taskExecutor = new PageTaskExecutor(this.page, {
      cacheId: opts?.cacheId,
    });
    this.reportFileName = reportFileName(opts?.testId || 'web');
  }

  appendExecutionDump(execution: ExecutionDump) {
    const currentDump = this.dump;
    currentDump.executions.push(execution);
  }

  dumpDataString() {
    // update dump info
    this.dump.groupName = this.opts.groupName!;
    this.dump.groupDescription = this.opts.groupDescription;
    return stringifyDumpData(this.dump);
  }

  writeOutActionDumps() {
    const { generateReport, autoPrintReportMsg } = this.opts;
    this.reportFile = writeLogFile({
      fileName: this.reportFileName!,
      fileExt: groupedActionDumpFileExt,
      fileContent: this.dumpDataString(),
      type: 'dump',
      generateReport,
    });

    if (generateReport && autoPrintReportMsg) {
      printReportMsg(this.reportFile);
    }
  }

  async aiAction(taskPrompt: string) {
    const { executor } = await this.taskExecutor.action(taskPrompt);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
  }

  async aiQuery(demand: any) {
    const { output, executor } = await this.taskExecutor.query(demand);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
    return output;
  }

  async aiAssert(assertion: string, msg?: string) {
    const { output, executor } = await this.taskExecutor.assert(assertion);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (!output?.pass) {
      const errMsg = msg || `Assertion failed: ${assertion}`;
      const reasonMsg = `Reason: ${output?.thought || '(no_reason)'}`;
      throw new Error(`${errMsg}\n${reasonMsg}`);
    }
  }

  async aiWaitFor(assertion: string, opt?: AgentWaitForOpt) {
    const { executor } = await this.taskExecutor.waitFor(assertion, {
      timeoutMs: opt?.timeoutMs || 15 * 1000,
      checkIntervalMs: opt?.checkIntervalMs || 3 * 1000,
      assertion,
    });
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
  }

  async ai(taskPrompt: string, type = 'action') {
    if (type === 'action') {
      return this.aiAction(taskPrompt);
    }
    if (type === 'query') {
      return this.aiQuery(taskPrompt);
    }

    if (type === 'assert') {
      return this.aiAssert(taskPrompt);
    }

    throw new Error(
      `Unknown type: ${type}, only support 'action', 'query', 'assert'`,
    );
  }
}
