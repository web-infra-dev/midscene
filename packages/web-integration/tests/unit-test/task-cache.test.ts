import { type LocateTask, type PlanTask, TaskCache } from '@/common/task-cache';
import type { WebUIContext } from '@/common/utils';
import type { WebElementInfo } from '@/web-element';
import type { AIElementParseResponse } from '@midscene/core';
import { beforeEach, describe, expect, it } from 'vitest';

describe('TaskCache', () => {
  let taskCache: TaskCache;
  let formalPageContext: WebUIContext;
  let pageContext: LocateTask['pageContext'];

  beforeEach(() => {
    taskCache = new TaskCache();
    pageContext = {
      url: 'https://example.com',
      size: { width: 1024, height: 768 },
    };
    formalPageContext = {
      ...pageContext,
      screenshotBase64: '',
      content: [{ id: 'element1' } as WebElementInfo], // 示例页面内容
    };
  });

  it('should return false if no cache is available', async () => {
    const cacheGroup = taskCache.getCacheGroupByPrompt('test prompt');
    const result = cacheGroup.readCache(
      formalPageContext,
      'plan',
      'test prompt',
    );
    expect(result).toBe(false);
  });

  it('should return false if the prompt does not match', async () => {
    taskCache.cache.aiTasks = [
      {
        prompt: 'different prompt',
        tasks: [
          {
            type: 'plan',
            prompt: 'different prompt',
            pageContext,
            response: { plans: [] },
          },
        ],
      },
    ];
    const cacheGroup = taskCache.getCacheGroupByPrompt('test prompt');
    const result = cacheGroup.readCache(
      formalPageContext,
      'plan',
      'test prompt',
    );
    expect(result).toBe(false);
  });

  it('should return false if the element cannot be found in the new context', async () => {
    taskCache.cache = {
      aiTasks: [
        {
          prompt: 'test prompt',
          tasks: [
            {
              type: 'locate',
              prompt: 'test prompt',
              pageContext,
              response: {
                elements: [{ id: 'element3' }],
              } as AIElementParseResponse,
            },
          ],
        },
      ],
    };
    const cacheGroup = taskCache.getCacheGroupByPrompt('test prompt');
    const result = cacheGroup.readCache(
      formalPageContext,
      'locate',
      'test prompt',
    );
    expect(result).toBe(false);
  });

  it('should return cached response if the conditions match', async () => {
    const cachedResponse = {
      plans: [{ type: 'Locate', thought: '', param: {} }],
    } as PlanTask['response'];
    taskCache.cache = {
      aiTasks: [
        {
          prompt: 'test prompt',
          tasks: [
            {
              type: 'plan',
              prompt: 'test prompt',
              pageContext,
              response: cachedResponse,
            },
          ],
        },
      ],
    };

    const cacheGroup = taskCache.getCacheGroupByPrompt('test prompt');
    const result = cacheGroup.readCache(
      formalPageContext,
      'plan',
      'test prompt',
    );
    expect(result).toEqual(cachedResponse);
  });

  it('should save cache correctly', () => {
    const cacheGroup = taskCache.getCacheGroupByPrompt('test prompt');
    const newCache: PlanTask = {
      type: 'plan',
      prompt: 'new prompt',
      pageContext,
      response: { plans: [{ type: 'Locate', thought: '', param: {} }] },
    };
    cacheGroup.saveCache(newCache);
    expect(taskCache.newCache.aiTasks[0].tasks).toContain(newCache);
  });

  it('should check page context equality correctly', () => {
    const isEqual = taskCache.pageContextEqual(pageContext, formalPageContext);
    expect(isEqual).toBe(true);

    const differentContext = {
      ...formalPageContext,
      size: { width: 800, height: 600 },
    };
    const isNotEqual = taskCache.pageContextEqual(
      pageContext,
      differentContext,
    );
    expect(isNotEqual).toBe(false);
  });

  it('should generate task cache correctly', () => {
    const generatedCache = taskCache.generateTaskCache();
    expect(generatedCache).toEqual(taskCache.newCache);
  });
});
