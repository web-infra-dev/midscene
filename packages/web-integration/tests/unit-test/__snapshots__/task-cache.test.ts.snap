// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCache > save and retrieve cache from file 1`] = `
[
  {
    "prompt": "test",
    "type": "plan",
    "yamlWorkflow": "test-yaml-workflow",
  },
  {
    "prompt": "test-locate",
    "type": "locate",
    "xpaths": [
      "test-xpath-1",
      "test-xpath-2",
    ],
  },
]
`;

exports[`TaskCache > save and retrieve cache from file 2`] = `
[
  {
    "prompt": "test",
    "type": "plan",
    "yamlWorkflow": "test-yaml-workflow",
  },
  {
    "prompt": "test-locate",
    "type": "locate",
    "xpaths": [
      "test-xpath-3",
      "test-xpath-4",
    ],
  },
]
`;

exports[`TaskCache > save and retrieve cache from file 3`] = `
"midsceneVersion: 0.16.11
cacheId: cacheId
caches:
  - type: plan
    prompt: test
    yamlWorkflow: test-yaml-workflow
  - type: locate
    prompt: test-locate
    xpaths:
      - test-xpath-3
      - test-xpath-4
"
`;

exports[`TaskCache > should create cache file 1`] = `
"midsceneVersion: 0.16.11
cacheId: cacheId
caches:
  - type: plan
    prompt: test
    yamlWorkflow: test
"
`;

exports[`TaskCache > update or append cache record - should not match cache added in same run 1`] = `
[
  {
    "prompt": "test-prompt",
    "type": "plan",
    "yamlWorkflow": "test-yaml-workflow",
  },
  {
    "prompt": "test-prompt",
    "type": "plan",
    "yamlWorkflow": "test-yaml-workflow-2",
  },
]
`;
